lib/snapshot.go

```diff
@@ -0,0 +1,1049 @@
+}
+
+
+// GetSnapshotChunk fetches a batch of records from the nodes DB that match the provided prefix and
+// have a key at least equal to the startKey lexicographically. The function will also fetch ancestral
+// records and combine them with the DB records so that the batch reflects an ancestral block.
+func (snap *Snapshot) GetSnapshotChunk(mainDb *badger.DB, prefix []byte, startKey []byte) (
+	_snapshotEntriesBatch []*DBEntry, _snapshotEntriesFilled bool, _concurrencyFault bool, _err error) {
+
+	// Check if we're flushing to the main db or to the ancestral records. If a flush is currently
+	// taking place, we will return a concurrencyFault error because the records are getting modified.
+	snap.SemaphoreLock.Lock()
+	ancestralDBSemaphoreBefore := snap.AncestralDBSemaphore
+	mainDBSemaphoreBefore := snap.MainDBSemaphore
+	snap.SemaphoreLock.Unlock()
+	if snap.isFlushing() {
+		return nil, false, true, nil
+	}
+
+	// This the list of fetched DB entries.
+	var snapshotEntriesBatch []*DBEntry
+
+	// Fetch the batch from main DB records with a batch size of about snap.BatchSize.
+	mainDbBatchEntries, mainDbFilled, err := DBIteratePrefixKeys(mainDb, prefix, startKey, SnapshotBatchSize)
+	if err != nil {
+		return nil, false, false, errors.Wrapf(err, "Snapshot.GetSnapshotChunk: Problem fetching main Db records: ")
+	}
+	// Fetch the batch from the ancestral DB records with a batch size of about snap.BatchSize.
+	ancestralDbBatchEntries, ancestralDbFilled, err := DBIteratePrefixKeys(snap.AncestralRecordsDb,
+		snap.GetAncestralRecordsKey(prefix), snap.GetAncestralRecordsKey(startKey), SnapshotBatchSize)
+	if err != nil {
+		return nil, false, false, errors.Wrapf(err, "Snapshot.GetSnapshotChunk: Problem fetching main Db records: ")
+	}
+
+	// To combine the main DB entries and the ancestral records DB entries, we iterate through the ancestral records and
+	// for each key we add all the main DB keys that are smaller than the currently processed key. The ancestral records
+	// entries have priority over the main DB entries, so whenever there are entries with the same key among the two DBs,
+	// we will only add the ancestral record entry to our snapshot batch. Also, the loop below might appear like O(n^2)
+	// but it's actually O(n) because the inside loop iterates at most O(n) times in total.
+
+	// Index to keep track of how many main DB entries we've already processed.
+	indexChunk := 0
+	for _, ancestralEntry := range ancestralDbBatchEntries {
+		dbEntry := snap.AncestralRecordToDBEntry(ancestralEntry)
+
+		for jj := indexChunk; jj < len(mainDbBatchEntries); {
+			if bytes.Compare(mainDbBatchEntries[jj].Key, dbEntry.Key) == -1 {
+					snapshotEntriesBatch = append(snapshotEntriesBatch, mainDbBatchEntries[jj])
+			} else if bytes.Compare(mainDbBatchEntries[jj].Key, dbEntry.Key) == 1 {
+				break
+			}
+			// if keys are equal we just skip
+			jj ++
+			indexChunk = jj
```

Comment:
```suggestion
			comparison := bytes.Compare(mainDbBatchEntries[jj].Key, dbEntry.Key)
			if comparison == -1 {
				snapshotEntriesBatch = append(snapshotEntriesBatch, mainDbBatchEntries[jj])
			} else if comparison == 1 {
				break
			}
			// if keys are equal we just skip
			jj ++
			indexChunk = jj
```
let's only call bytes.Compare once if we can help it.

Reply:
good one 👍